### ISSUE TITLE
How do we generate Matterport 3D data?

### ISSUE THREAD

**Issue by etri-jeon0123** [facebookresearch/habitat-lab#1457]
Habitat-Lab: v0.2.4
Habitat-Sim: v0.2.4

Currently, we are going to use Matterport 3D equipment to shoot a new environment and use the filmed location as input.

In the cloud operated by the Matterport company, 
we can receive **.mlt and .obj** files, and I understand that we need to convert them into **.glb** files for use with Habitat.

However, there is an issue that cannot be used immediately because the composition of Matterport3D (MP3D) data is not just .glb.

The MP3D data is configured as follows.

- filename.glb
- filename.house
- filename.navmesh
- filename_semantic.ply
- mp3d.scene_dataset_config.json

Therefore, we can't make it except for the .glb file, so I wonder how we can create it.

**Comment by etri-jeon0123** [comment]
> Hey @etri-jeon0123,
> 
> What you are describing here is creating a [SceneDataset](https://aihabitat.org/docs/habitat-sim/attributesJSON.html): the ecosystem of assets and configuration and metadata files necessary to utilize the full set of Habitat features.
> 
> Of course you can do basic embodied tasks such as point nav with only the glb file. The other files have purposes and origins as follows:
> 
> * `.house` - a Semantic Scene Descriptor (SSD) file. This contains a mapping from instance ids to semantic classes (e.g. instance 13 may be a table) as well as optionally defining object and region bounding boxes.
> * `.navmesh` - the navigation constraints for the scene. This is computed automatically if not provided, or can be saved and loaded. See the [nav API](https://aihabitat.org/docs/habitat-sim/habitat_sim.nav.html) for more details.
> * `semantic.ply` - in MP3D dataset these are additional assets (mesh + color material) containing per-vertex semantic annotations. As of [HM3D](https://aihabitat.org/datasets/hm3d-semantics/) you can also provide these annotations in texture format instead.
>   
>   * For example, all vertices beloning to a table may have a color mapping to integer 13 as defined in the .house file.
>   * You will need to manually annotate these for your new scene assets. One option is to paint the textures in a modeling software such as Blender. Paint each instance with a unique color and map those colors to integer ids in a semantic.txt file. See HM3D for examples.
> * `.scene_dataset_config.json` - this registry file simply identifies the assets and configs available in a full scene dataset. Use MP3D or HM3D as an example for how to setup the directory structure and aggregate the results.
> 
> Everything but the semantic annotation and mapping is pretty easy to put together following existing examples. For that part, if necessary, I suggest you look into the assets and semantic txt files from HM3D as that process is likely easier and results in better annotation accuracy than the per-vertex colors embedded in the MP3D dataset.

Thank you for your valuable reply.
But the question here is, how can we create the files described above.

In other words, how can we create **.house/.navmesh/.semantic.ply** on our own?

In the case of file **config.json**, I think we can make it by referring to other files, but I don't know the exact method of other files.

Therefore, can you explain how to create **.house, .navmesh and .semantic.ply files** for the generated .glb files?

**Comment by aclegg3** [comment]
@etri-jeon0123, happy to provide some more details here. 

The .navmesh file will be created automatically by the simulator and if you want to customize it see the API I linked above. 

The most challenging part here is the annotation of the semantic asset (.house and .semantic.ply). 

First, **I suggest you follow the format of HM3D Semantics rather than MP3D and instead generate a `.semantic.glb` and a `.semantic.txt` file.** Please review our arxiv report (https://arxiv.org/abs/2210.05633) for many details regarding this process and the file contents.

TL;DR: 

The `.semantic.txt` file is hand-authored and each line contains metadata corresponding to a single object instance (e.g. a lamp) linking the semantic texture color to an integer id and semantic label. This is read by the engine to understand the colors embedded in the texture files from `.semantic.glb`. The glb file is the same glb as your original asset, but instead of RGB textures from the scan, you re-color those textures with unique per-instance colors mapped in the text file. 

- For example scene GLAQ4DNUx5U, the first line is `1,C9C373,"door",0`. This tells us that instance 1 with hex color `C9C373` is a door in region 0.

For the existing dataset, this was done by resetting the texture images to black and then painting new colors from a chosen palette on to the 3D geometry. This is an artistic process and can be labor intensive if the contents of the scene are complex and detailed.

- In the example, an artist painted the color `C9C373` onto a door in the scene geometry and then logged it in the text file.

**Comment by etri-jeon0123** [comment]
> @etri-jeon0123, happy to provide some more details here.
> 
> The .navmesh file will be created automatically by the simulator and if you want to customize it see the API I linked above.
> 
> The most challenging part here is the annotation of the semantic asset (.house and .semantic.ply).
> 
> First, **I suggest you follow the format of HM3D Semantics rather than MP3D and instead generate a `.semantic.glb` and a `.semantic.txt` file.** Please review our arxiv report (https://arxiv.org/abs/2210.05633) for many details regarding this process and the file contents.
> 
> TL;DR:
> 
> The `.semantic.txt` file is hand-authored and each line contains metadata corresponding to a single object instance (e.g. a lamp) linking the semantic texture color to an integer id and semantic label. This is read by the engine to understand the colors embedded in the texture files from `.semantic.glb`. The glb file is the same glb as your original asset, but instead of RGB textures from the scan, you re-color those textures with unique per-instance colors mapped in the text file.
> 
> * For example scene GLAQ4DNUx5U, the first line is `1,C9C373,"door",0`. This tells us that instance 1 with hex color `C9C373` is a door in region 0.
> 
> For the existing dataset, this was done by resetting the texture images to black and then painting new colors from a chosen palette on to the 3D geometry. This is an artistic process and can be labor intensive if the contents of the scene are complex and detailed.
> 
> * In the example, an artist painted the color `C9C373` onto a door in the scene geometry and then logged it in the text file.

That's a very precious answer.
Immediately, we changed the data to review the paper with respect to HM3D and even complete the test.

However, we still have questions about organizing our own data.

For our environment, we confirmed that we have to make **`semantic.glb`** & **`semantic.txt`**.

`[Question 1]`
And, to make semantic.glb, the existing basis.glb should be used in the semantic segmentation method, but your paper said it was done manually.
**`On average, each scene consists of 661 objects from 106 categories. This dataset is the result of over 14,200 hours of human effort for annotation and verification by 20+ annotators.`**

If we want to make semantic.glb, is it right to convert it using a model like ScanNet as shown in Table 1?
![image](https://github.com/facebookresearch/habitat-lab/assets/121843576/ad4a0b2c-ec18-4707-a5b3-8f8c4236289b)

`[Question 2]`
Then, as a second question, where can I find the **`color, object, and region`** (`1,C9C373,"door",0`) information in semantic.txt? Even if we try to make a label manually, there are no previously defined color, object, region information, can you share it?

With your help, we are working on a lot of handrails, but there are still some hurdles left, so I'm asking for your help.

**Comment by aclegg3** [comment]
> And, to make semantic.glb, the existing basis.glb should be used in the semantic segmentation method, but your paper said it was done manually.

Yes, these annotations were done manually. We are not aware of an existing automated approach to semantic annotations with high accuracy. We suggest painting the semantic annotations on the assets in a modeling software such as Blender.

> where can I find the color, object, and region

Again, these are chosen by the annotator during the painting process.
For each object instance, the annotator will choose an unused color, paint the corresponding geometry, and then record the color and the semantic class in the txt file. 
For example, [this file](https://github.com/facebookresearch/habitat-sim/blob/main/data/hm3d_semantics/hm3dsem_category_mappings.tsv) maps the annotated classes in HM3D to a category and canonical  mpcat40 lexicon from[ this paper](https://niessner.github.io/Matterport/).

In summary:
- The semantic annotation process is manual. After scanning a space, you will need to paint the textures and then map them to a set of categories.
- You can choose any categories you want. You are welcome to use the ones we've already defined in the linked mappings file or your own.
- You can choose any colors you want. Each color must be used only once for each scene.
- Region information is also manually defined. Typically a region is a single room, but Habitat does not use this annotation for any simulation logic by default, so you can use this field however you like.

**Comment by etri-jeon0123** [comment]
> > And, to make semantic.glb, the existing basis.glb should be used in the semantic segmentation method, but your paper said it was done manually.
> 
> Yes, these annotations were done manually. We are not aware of an existing automated approach to semantic annotations with high accuracy. We suggest painting the semantic annotations on the assets in a modeling software such as Blender.
> 
> > where can I find the color, object, and region
> 
> Again, these are chosen by the annotator during the painting process. For each object instance, the annotator will choose an unused color, paint the corresponding geometry, and then record the color and the semantic class in the txt file. For example, [this file](https://github.com/facebookresearch/habitat-sim/blob/main/data/hm3d_semantics/hm3dsem_category_mappings.tsv) maps the annotated classes in HM3D to a category and canonical mpcat40 lexicon from[ this paper](https://niessner.github.io/Matterport/).
> 
> In summary:
> 
> * The semantic annotation process is manual. After scanning a space, you will need to paint the textures and then map them to a set of categories.
> * You can choose any categories you want. You are welcome to use the ones we've already defined in the linked mappings file or your own.
> * You can choose any colors you want. Each color must be used only once for each scene.
> * Region information is also manually defined. Typically a region is a single room, but Habitat does not use this annotation for any simulation logic by default, so you can use this field however you like.

When working on modeling software such as Blender, **did you import and work on 3D modeling file formats such as .GLB? Or did you convert the .GLB file into a 2D image to work?**

We are glad to know briefly about the composition of the data and what to do with your help. However, in order to work on semantic segmentation of 3D space in detail, we need more detailed information, so we ask for your help.

Perhaps, we think that many things will be solved when we learn how to perform semantic segmentation tasks on modeling software such as Blender.

Thank you for your help in advance.

**Comment by aclegg3** [comment]
> did you import and work on 3D modeling file formats such as .GLB? Or did you convert the .GLB file into a 2D image to work?

We import the GLB file into Blender, replace the existing texture images with empty (black) images, and then fill new content on the 2D images by directly painting on the 3D geometry.
- For example, see this video: https://youtu.be/WjS_zNQNVlw?t=177

You can pick a color for each object instance or feature you want to segment and then paint it into the geometry.

**Comment by Rendok** [comment]
Hi @aclegg3,

I'm trying to make my custom scene. I have <scene>.obj and I converted it into <scene>.glb. But when I try to make .navmesh via `build/utils/datatool/Datatool create_navmesh <scene>.glb <scene>.navmesh`. I'm getting an error: `ESP_CHECK failed: esp::logging::LoggingContext: No current logging context.`

What is the proper way of doing it?

**Comment by aclegg3** [comment]
Hey @Rendok 

You don't need the Datatool to create the NavMesh. You can set the desired NavMeshSettings and provide them to [SimulatorConfiguration.navmesh_settings](https://aihabitat.org/docs/habitat-sim/habitat_sim.sim.SimulatorConfiguration.html#navmesh_settings) upon initialization to recompute the navmesh at runtime. You can also take a look at our [ECCV 2020 tutorial scripts](https://github.com/facebookresearch/habitat-sim/blob/435d942386b1ce774e03d1fd1dafa7a54b33c888/examples/tutorials/nb_python/ECCV_2020_Navigation.py#L743) to get some example code for using the NavMesh as well as recomputing and saving/loading it.
